### PR TITLE
QPID-8454 Add prometheus metrics

### PR DESCRIPTION
**Pull request by Dedeepya-T**
The Pull Request adds the Qpid Broker integration with Prometheus
* a new broker plugin modules qpid-prometheus-exporter is added to collect Qpid broker statistics as Prometheus metrics
* QpidCollector implements the Prometheus Client Collector interface
* a new Plugin interface ContentFactory is added into the broker-core
* an implementation of ContentFactory is for Prometheus metrics
* a new servlet is added to serve content generated with ContentFactory
* Prometheus servlets are mapped under /metrics and /metrics/*
* /metrics serves broker metrics
* /metrics/<virtual host node name>/<virtual host name> serves Virtual host metrics

### CODE REVIEW

File: broker-core/src/main/java/org/apache/qpid/server/model/port/AmqpPort.java [hunk 1]
```diff
@@ -159,12 +159,16 @@
...
+            metricName = "open_connections_total")
     int getConnectionCount();
 
-    @ManagedStatistic(statisticType = StatisticType.CUMULATIVE, units = StatisticUnit.COUNT, label = "Total Connections",
-            description = "Total number of connections made through this port since broker startup")
+    @ManagedStatistic(statisticType = StatisticType.CUMULATIVE, units = StatisticUnit.COUNT,
+            label = "Total Connections",
+            description = "Total number of connections made through this port since broker startup",
+            metricName = "total_connections_count")
     long getTotalConnectionCount();
```
Comment (alex-rufous) [inline broker-core/src/main/java/org/apache/qpid/server/model/port/AmqpPort.java:172]: As total is used for gauges in prometheus, this metric can be called overall_connections_count or aggregate_connection_count

File: broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java [hunk 2]
```diff
@@ -215,81 +215,81 @@
...
 
     @SuppressWarnings("unused")
     @ManagedStatistic(statisticType = StatisticType.POINT_IN_TIME, units = StatisticUnit.COUNT, label = "Connections",
-                      description = "Current number of messaging connections made to this virtualhost.")
+                      description = "Current number of messaging connections made to this virtualhost.", metricName = "connections_total")
     long getConnectionCount();
 
     @ManagedStatistic(statisticType = StatisticType.CUMULATIVE, units = StatisticUnit.COUNT, label = "Total Connections",
-            description = "Total number of messaging connections made to this virtualhost since broker startup")
+            description = "Total number of messaging connections made to this virtualhost since broker startup", metricName = "connections_since_startup_count")
```
Comment (alex-rufous) [inline broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java:232]: The metric name is inconsistent with the same metric name on amqp port . Possibly, overall_connections_count or aggregate_connection_count could be a better name for the metric.

File: broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java [hunk 3]
```diff
@@ -215,81 +215,81 @@
...
     @ManagedStatistic(statisticType = StatisticType.CUMULATIVE, units = StatisticUnit.MESSAGES,
             label = "Transacted Outbound",
-            description = "Total number of messages received by this virtualhost within a transaction.")
+            description = "Total number of messages received by this virtualhost within a transaction.", metricName = "outbound_transacted_messages_count")
     long getTransactedMessagesOut();
 
     @SuppressWarnings("unused")
     @ManagedStatistic(statisticType = StatisticType.POINT_IN_TIME, units = StatisticUnit.BYTES, label = "Queue Depth",
-            description = "Current size of all messages enqueued by this virtualhost.")
+            description = "Current size of all messages enqueued by this virtualhost.", metricName = "total_queue_depth_bytes_total")
```
Comment (alex-rufous) [inline broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java:270]: The Virtual Host metric total_queue_depth_bytes_total contains "total"  twice. It should be renamed into queue_depth_bytes_total

File: broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java [hunk 4]
```diff
@@ -215,81 +215,81 @@
...
     long getTotalDepthOfQueuesBytes();
 
     @SuppressWarnings("unused")
     @ManagedStatistic(statisticType = StatisticType.POINT_IN_TIME, units = StatisticUnit.MESSAGES, label = "Queue Depth",
-                      description = "Current number of messages enqueued by this virtualhost.")
+                      description = "Current number of messages enqueued by this virtualhost.", metricName = "total_queue_depth_messages_total")
```
Comment (alex-rufous) [inline broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java:275]: Total is included twice into metric name. It should be queue_depth_messages_total

File: broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java [hunk 5]
```diff
@@ -215,81 +215,81 @@
...
-                      description = "Total Number of Bytes Evacuated from Memory Due to Flow to Disk.")
+                      description = "Total Number of Bytes Evacuated from Memory Due to Flow to Disk.", metricName = "evacuated_messages_bytes_count")
     long getBytesEvacuatedFromMemory();
 
     @SuppressWarnings("unused")
     @ManagedStatistic(statisticType = StatisticType.POINT_IN_TIME,
             units = StatisticUnit.BYTES,
             label = "Maximum recorded size of inbound messages",
-            description = "Maximum size of message published into the Virtual Host since start-up.")
+            description = "Maximum size of message published into the Virtual Host since start-up.", metricName = "high_water_mark_inbound_messages_total")
```
Comment (alex-rufous) [inline broker-core/src/main/java/org/apache/qpid/server/virtualhost/QueueManagingVirtualHost.java:292]: Total needs to be removed from water mark metric. The name inbound_message_size_high_watermark would work better here

File: broker-plugins/prometheus-exporter/src/main/java/org/apache/qpid/server/prometheus/PrometheusContentFactory.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        final String[] includeDisabledValues = filter.get("includeDisabled");
+        final boolean includeDisabled = includeDisabledValues!= null && includeDisabledValues.length == 1 && Boolean.parseBoolean(includeDisabledValues[0]);
+        final String[] includedMetricNames = filter.get("name[]");
+
+        final IncludeMetricPredicate metricIncludeFilter =
+                new IncludeMetricPredicate(includedMetricNames == null || includedMetricNames.length == 0
+                                                   ? Collections.emptySet()
+                                                   : new HashSet<>(Arrays.asList(includedMetricNames)));
+        final QpidCollector qpidCollector = new QpidCollector(object,
+                                                              new IncludeDisabledStatisticPredicate(includeDisabled),
```
Comment (alex-rufous) [inline broker-plugins/prometheus-exporter/src/main/java/org/apache/qpid/server/prometheus/PrometheusContentFactory.java:55]: The flag for inclusion of disabled metrics can be set via context variable. Potentially,  the factory can check whether context variable like qpid.metrics.includeDisabled is set to true on the root object.

File: broker-plugins/prometheus-exporter/src/main/java/org/apache/qpid/server/prometheus/QpidCollector.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    private static String generateMetricSuffix(final ConfiguredObjectStatistic<?, ?> statistics, String typeSuffix)
+    {
+        String suffix = "";
+        if (!statistics.getName().toLowerCase().contains(typeSuffix)
+            && statistics.getUnits() != StatisticUnit.ABSOLUTE_TIME
+            && statistics.getUnits() != StatisticUnit.TIME_DURATION)
+        {
+            if (statistics.getUnits() == StatisticUnit.MESSAGES || statistics.getUnits() == StatisticUnit.BYTES)
+            {
+                suffix = "_" + statistics.getUnits().toString() + "s";
```
Comment (alex-rufous) [inline broker-plugins/prometheus-exporter/src/main/java/org/apache/qpid/server/prometheus/QpidCollector.java:245]: The metric name functionality can be improved to check whether metric name already ends with unit names. When units are there, skip the unit addition to the suffix

**Review by alex-rufous — CHANGES_REQUESTED** [review]
Hi Dedeepya,
Please rebase your repo branch against github master,  address the review comments and resubmit the pull request.

Thanks,
Alex

**Comment by Dedeepya-T** [comment]
Hi Alex, Thanks for taking time to review the code. I have addressed all the above comments and ready for review.

**Pull request closed without merging**